packages/pigeon/lib/cpp_generator.dart

```diff
@@ -0,0 +1,683 @@
+      });
+      indent.writeScoped('protected:', '', () {
+        indent.writeln(
+            'flutter::EncodableValue ReadValueOfType(uint8_t type, flutter::ByteStreamReader* stream) const;');
+      });
+    }
+  });
+}
+
+void _writeCodecSource(Indent indent, Api api, Root root) {
+  final String codecName = _getCodecName(api);
+  indent.writeln('$codecName::$codecName() {}');
+  if (getCodecClasses(api, root).isNotEmpty) {
+    indent.write(
+        'flutter::EncodableValue $codecName::ReadValueOfType(uint8_t type, flutter::ByteStreamReader* stream) const');
+    indent.scoped('{', '}', () {
+      indent.write('switch (type) ');
+      indent.scoped('{', '}', () {
+        for (final EnumeratedClass customClass in getCodecClasses(api, root)) {
+          indent.write('case ${customClass.enumeration}: ');
+          indent.writeScoped('', '', () {
+            indent.writeln(
+                'return flutter::CustomEncodableValue(${customClass.name}(std::get<flutter::EncodableMap>(ReadValue(stream))));');
+          });
+        }
+        indent.write('default:');
+        indent.writeScoped('', '', () {
+          indent.writeln(
+              'return flutter::StandardCodecSerializer::ReadValueOfType(type, stream);');
+        });
+      });
+    });
+    indent.write(
+        'void $codecName::WriteValue(const flutter::EncodableValue& value, flutter::ByteStreamWriter* stream) const');
+    indent.writeScoped('{', '}', () {
+      indent.write(
+          'if(const flutter::CustomEncodableValue* customValue = std::get_if<flutter::CustomEncodableValue>(&value))');
+      indent.scoped('{', '} else ', () {
+        for (final EnumeratedClass customClass in getCodecClasses(api, root)) {
+          indent
+              .write('if(customValue->type() == typeid(${customClass.name}))');
+          indent.scoped('{', '} else ', () {
+            indent.writeln('stream->WriteByte(${customClass.enumeration});');
+            indent.writeln(
+                'WriteValue(std::any_cast<${customClass.name}>(*customValue).ToEncodableMap(), stream);');
+          });
+        }
+        indent.scoped('{', '}', () {
+          indent.writeln(
+              'flutter::StandardCodecSerializer::WriteValue(value, stream);');
+        });
+      });
+      indent.scoped('{', '}', () {
+        indent.writeln(
+            'flutter::StandardCodecSerializer::WriteValue(value, stream);');
+      });
+    });
+  }
+}
+
+void _writeHostApiHeader(Indent indent, Api api) {
+  assert(api.location == ApiLocation.host);
+
+  indent.writeln(
+      '/** Generated class from Pigeon that represents a handler of messages from Flutter.*/');
+  indent.write('class ${api.name} ');
+  indent.scoped('{', '};', () {
+    indent.scoped('public:', '', () {
+      for (final Method method in api.methods) {
+        final String returnType = method.isAsynchronous
+            ? 'void'
+            : _cppTypeForDartType(method.returnType);
```

Comment:
I'm concerned about the structure here; you've eliminated the ability to do the equivalent of `result->Error` in a non-Pigeon plugin, which is important functionality. AFAIK the only way to return an error would be to throw which a) we should absolutely not be encouraging/requiring people to do as a normal error handling mechanism in C++, and b) provides less information than a usual method channel error (code + description + arbitrary details).

We need to decide on a structure that allows for errors. Maybe that's something like a generated `ErrorOr<SomeReturnType>` to use as the return value?

Reply:
I did a similar implementation to the Java one. The one info that I'm not wrapping is the stack trace, but if the developer throws an exception, it will be handled and wrapped properly. The iOS implementation has a FlutterError type, so we could generate a generic equivalent of that, like your ErrorOr<T>.

Reply:
Ok, this took me a while, but this seems to be working quite well.